RequestManager.py

```diff
@@ -213,25 +213,55 @@ def upload_indicators(self, parsed_indicators):
             request_body = {"sourcesystem": "MISP", "value": parsed_indicators[:config.ms_max_indicators_request]}
-            response = requests.post(request_url, headers=self.headers, json=request_body)
-            if response.status_code == 200:
-                if "errors" in response.json() and len(response.json()["errors"]) > 0:
-                    if config.sentinel_write_response:
-                        json_formatted_str = json.dumps(response.json(), indent=4)
-                        with open("sentinel_response.txt", "a") as fp:
-                            fp.write(json_formatted_str)                    
-                    self.logger.error("Error when submitting indicators - error string received from Sentinel. {}".format(response.text))
+
+            # Setting result retry as true to enter the loop
+            result = {"retry": True, "breakRun": False}
+
+            while result.get("retry", True):
+                response = requests.post(request_url, headers=self.headers, json=request_body)
+                requests_number += 1
+                result = self.handle_response_codes(response, safe_margin, requests_number, request_body, parsed_indicators)
+                # If retry is true, retry the request
+                # If breakRun is true, break out of the loop
+                if result.get("breakRun", True):
                     break
-                else:
-                    parsed_indicators = parsed_indicators[config.ms_max_indicators_request:]
-                    self.logger.info("Indicators sent - request number: {} / indicators: {} / remaining: {}".format(requests_number, len(request_body["value"]), len(parsed_indicators)))
-                    requests_number += 1
-            else:
-                self.logger.error("Error when submitting indicators. Non HTTP-200 response. {}".format(response.text))
-                if response.status_code == 500:
-                    self.logger.error("request_body={}".format(request_body))
-                break
 
+    def handle_response_codes(self, response, safe_margin, requests_number, request_body, parsed_indicators):
+        self.logger.debug(response)
+        status_code = response.status_code
+        result = {}
+        switcher = {
+            429: lambda: self.handle_rate_limit_exceeded(response, safe_margin),
+            200: lambda: self.handle_success_response(response, request_body, parsed_indicators, requests_number),
+        }
+        switcher.get(status_code, lambda: self.handle_error_response(response))()
+        self.logger.debug(result)
+        return result
+
+    def handle_rate_limit_exceeded(self, response, safe_margin):
+        error_message = response.json()["message"]
+        retry_after = int(error_message.split()[-2])
+        self.logger.warning(f"Rate limit exceeded. Retrying after {retry_after} seconds.")
+        time.sleep(retry_after + safe_margin)
+        return {"retry": True, "breakRun": False}
```

Comment:
Could add a panic value that if retry exceeds >10 attempts kill it

Reply:
On the other hand, this should only run when a 'specific' (429) code response is there. So I don't think there's that risk here.